src/runtime-rs/crates/hypervisor/src/device/driver/vfio.rs

```diff
@@ -547,6 +589,31 @@ impl PCIeDevice for VfioDevice {
             return Ok(());
         }
 
+        // handle port devices
+        if !self.allocated {
+            let port_type = pcie_topo.get_pcie_port();
+            info!(sl!(), "XXXport type: {:?}", port_type);
+            let avail_port = match port_type {
+                PCIePort::NoPort => return Ok(()),
+                PCIePort::InvalidPort => {
+                    return Err(anyhow!("You can not set both RootPort|SwitchPort"))
```

Comment:
Maybe a dumb question but could you explain why a PCIe topology cannot have both root & switch ports?

Reply:
I think I should clarify this statement about `RootPort` and `SwitchPort`, they're both port types defined in kata runtime, which are not totally equal to such definitions in Qemu but they have some relationships. `RootPort` is defined in kata runtime which only contains [pcie-root-port in qemu cmdline](https://github.com/qemu/qemu/blob/master/docs/pcie.txt#L117). While `SwitchPort` contains `pcie-root-port and upstream/downstream`, both of which comes from the [guideline of qemu](https://github.com/qemu/qemu/blob/master/docs/pcie.txt#L121). And I just make it alignment with kata-runtime